src/Handler/Search.hs

```diff
@@ -21,26 +24,87 @@ import qualified Language.PureScript as P
 
 import qualified XMLArrows
 
+resultsPerPage :: Int
+resultsPerPage = 50
+
+maxPages :: Int
+maxPages = 10
+
 getSearchR :: Handler TypedContent
 getSearchR = do
   mquery <- lookupGetParam "q"
+  page <- lookupCurrentPage
+
   case mquery of
     Nothing -> redirect HomeR
     Just query -> do
-      results <- case tryParseType query of
-        Just ty | not (isSimpleType ty) -> searchForType ty
-        _ -> searchForName (toLower query)
+      (results, hasMore) <- case tryParseType query of
+        Just ty | not (isSimpleType ty) -> searchForType page ty
+        _ -> searchForName page (toLower query)
       selectRep $ do
-        provideRep (htmlOutput query results)
-        provideRep (jsonOutput results)
+        provideRep (htmlOutput query results page hasMore)
+        provideRep (jsonOutput results page hasMore)
   where
-    htmlOutput :: Text -> [SearchResult] -> Handler Html
-    htmlOutput query results = do
+    getPageLinkRenderer :: Handler (Maybe (Int -> Text))
+    getPageLinkRenderer = runMaybeT $ do
+      cr <- MaybeT getCurrentRoute
+      MaybeT $ do
+        urp <- getUrlRenderParams
+        getParams <- Map.fromList . reqGetParams <$> getRequest
+        return $ Just $ \p ->
+          urp cr $ Map.toList
+                 $ Map.insert "page" (pack $ show p) getParams
+
+    mkPaginationW :: Int -> Bool -> Handler (Maybe Widget)
+    mkPaginationW page hasMore = getPageLinkRenderer >>= \mr -> case mr of
+        Nothing -> return Nothing
+        Just mkPageLink ->
+          let allPages = [1..page]
+              mmNextPageLink = if hasMore
+                                  then Just $ if page >= maxPages
```

Comment:
I think for a one off job like this the cognitive burden of `Maybe (Maybe a)` is less than introducing another ADT the reader of the code has to keep in memory (and then forget about after realizing how irrelevant it is)